File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*,Node*> oldcopy;
        oldcopy[nullptr] = nullptr;

        Node* curr =  head;
        while (curr != nullptr){
            oldcopy[curr] = new Node(curr->val);
            curr = curr->next;
        }

        curr = head;
        while(curr != nullptr){
            Node* copy = oldcopy[curr];
            copy->next = oldcopy[curr->next];
            copy->random = oldcopy[curr->random];
            curr = curr->next;
        }

        return oldcopy[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        HashMap <Node, Node> oldcopy = new HashMap<>();
        oldcopy.put(null,null);

        Node curr =  head;
        while (curr != null){
            oldcopy.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        curr = head;
        while(curr != null){
            Node copy = oldcopy.get(curr);
            copy.next = oldcopy.get(curr.next);
            copy.random = oldcopy.get(curr.random);
            curr = curr.next;
        }

        return oldcopy.get(head);
    }
}